### DISCUSSION TITLE
Problem in filtering matches for a dating app.

### DISCUSSION THREAD

**Discussion by GauravCalidig** [supabase/supabase#8455] [Questions] [answered]
I am building a dating app where users can swipe right and left to like and unlike respectively.
![image](https://user-images.githubusercontent.com/90178033/185282074-835ed596-7eed-433d-ba74-045e4b4c8c50.png)
This is my table, I am appending `likedUserIds` and `unlikedUserIds` 

A match happens when user 1 likes user 2 and vice versa.

I want to get the matches using a single supabase query but not able to do so.

In layman terms what i need is first we get all likedUserIds for loggedIn user, then a for loop on all ids to check if the liked user also likes the logged in user.

the above approach is not a feasible solution that's why asking if there is a better way to achieve this

**Comment by GaryAustin1** [comment, marked as answer]
Not commenting on best way versus what you have.  An rpc call to a Postgres function could get it done in one API call.

Or with two API calls you can get thisUsers likes in an arrayOfLikeIds in first call.  Then in next call  you can use:
```
.in('id',arrayOfLikeIds) //finds thisPerson's matches
.contains('liked',[thisUsersId]) //and if thisPerson is also in one of the OtherPerson's likes
```

Not tested... and late.

**Reply by GauravCalidig** [reply]
Thanks for the reply @GaryAustin1 , I am new to rpc calls can you share an example of how to approach this with rpc?

**Reply by GauravCalidig** [reply]
Since the likes can be as many as 100k I think it's better if I approach this with rpc.

**Reply by GaryAustin1** [reply]
Yikes!  Or like I said at the beginning... the best way...  An array attached to each user with 100k values may not be the way to go versus a table with match pairs.  100K per user array is beyond my skill set though.

An API rpc call is just a way to call a Postgres function and run SQL. 

I'm not qualified to give you performance advice though, if an array approach would even work out versus some sort of like table with all matches.  So either Google for something similar with Postgres or SQL, or see if someone else comes along.

Edit:  The only performant way I can think of to handle this is to have a match table.
When a user1 swipes to add user2, you add user2 to user1's array and then check user2's array to see if user1 is in it.  
If so you add user1, user2 AND user2,user1 to the match table.

Then easy to find matches for user1, just search for user1 in in first column of the match table for user1.

When someone drops a user from their like list, you do the opposite,
Update the user1 like array - user2, and then delete user1,user2 AND user2,user1 from the match array if the pair exists....

This probably can be simplified with just a single pair instead of both reversed pairs.

**Reply by GauravCalidig** [reply]
Got it, Let me try it your way then, Thanks a lot for the reply.
